### DISCUSSION TITLE
Type hinting for type conversions in validators

### DISCUSSION THREAD

**Discussion by Samreay** [pydantic/pydantic#3997] [General]
Hi team, my use case for Pydantic is to validate JSON input in FastAPI and turn the input json into some internal cython classes for use internally. For purpose of this, I've come up with a simpler but analogous use case, which is how to properly define types when your validators change type.

For example:

```python
from pydantic import BaseModel, validator
import numpy as np


class FindMean(BaseModel):
    x: list[float]

    @validator("x")
    @classmethod
    def validate_x(cls, x: list[float]) -> np.ndarray:
        return np.array(x)


if __name__ == "__main__":
    example = FindMean(x=[1, 2, 3])
    print(example.x)
    print(type(example.x))

    tmp = example.x
```
 The type printed out is the correct numpy array, however, using mypy, it will think that `tmp` is `list[float]`, and naively, a user will probably think this too when reading the class.

What is the recommended way to handle transformed data with type annotations in pydantic to resolve these two issues and remove potential confusion? I was thinking one could possibly do `pre=True` in the validator and then change the annotation in the class itself, but then this makes the input data structure opaque and puts the burden of json object conversion onto the user.

Right now, I have been creating custom `to_model()` functions which do things like the numpy or pandas dataframe conversion when requested, but the downside of doing this is loosing the nice automatic failure to create the model if the conversion fails, which is a handy feature to preserve.

**Comment by samuelcolvin** [comment]
There's no `tmp` in your example, so hard to say.

If you're refering to `x` the variable inside `validate_x` it will have type `list[float]` provided you don't use `pre=True`.

**Comment by Samreay** [comment]
`tmp` is the last line of the example.

The general point was asking for recommendations/thoughts on how we should be type hinting when our validator deliberately changes the type of the input and keep things mypy consistent.

For example:

* a list of floats gets validated and turned into a numpy array
* a transport layer object gets turned into a non-transport class
* json input gets turned into a cython class

If this isn't something supported by Pydantic, not an issue, I can construct an interim class, but it seems like splitting the responsibilities of validating / transforming certain fields into two separate classes is not an ideal solution. I could also just turn off mypy type hinting here as well, because lord knows getting everything to work with mypy is a mammoth undertaking!

**Reply by samuelcolvin** [reply]
Well the validator would not return a different type.

What you're doing breaks type annotations, you can use cast, or "type ignore", or not break type hints.

Your effectively asking " having crashed my car, how should I drive it?".

**Reply by Samreay** [reply]
Alright, will create two separate classes, that was all I was asking. Thanks for confirming.

**Reply by samuelcolvin** [reply]
no problem, sorry for the delay.

**Reply by Samreay** [reply]
No worries mate. Thanks for all the hard work you put in :)

**Comment by smoshkovits-dn** [comment]
I have a question in the same area, but in my case the conversion is done after the model is fully populated, so the type of num can be both `int` or `TwoNumbers`. 

When I type `print(one_number_int)` In the auto-complete menu of PyCharm I get suggestions for both types, which is expected, so:

My question is: **Is there a way that I can hide `TwoNumbers`, in such a way that auto complete suggests only `int` related suggestions?** 

![pydantic_example](https://user-images.githubusercontent.com/80193155/183417748-c2280326-67c1-4b8d-86b1-42a23996d443.png)


```
from typing import Union, List

from pydantic import BaseModel, validator
from typing_extensions import Annotated


class TwoNumbers(BaseModel):
    num1: int = 23
    num2: int = 100

    def sum_nums(self) -> int:
        return self.num1 + self.num2

class OneNumber(BaseModel):
    num: Union[int, TwoNumbers]



if __name__ == "__main__":
    two_numbers: TwoNumbers = TwoNumbers()
    one_number_from_class: OneNumber = OneNumber(num=two_numbers)
    # print(one_number_from_class.num.)  # When trying to autocomplete here - Getting suggestions for both int and TwoNumbers
    one_number_from_class.num = one_number_from_class.num.sum_nums()
    one_number_int: OneNumber = OneNumber(num=5555)

    print(one_number_from_class.num)
    print(one_number_int)from pydantic import BaseModel, validator
```

**Reply by samuelcolvin** [reply]
Probably a question for the [pydantic pycharm plugin](https://github.com/koxudaxi/pydantic-pycharm-plugin)

**Reply by smoshkovits-dn** [reply]
Thanks.

**Comment by schodge** [comment]
Pardon me for pulling up an old thread, but I think this is relevant to my question.

We use UUIDs with SqlAlchemy, but have some problems with how they display, how they serialize, etc. Using Pydantic 1.x, perhaps this would be possible more cleanly with `Annotated` as a type in Pydantic 2.

Anyway, I want a UUID that:
- given None, generates a random uuid4
- given the string '', creates the special UUID(0000...)
- given the string form of a UUID, creates the UUID of it
- has some options around serialization and display

I've wound up with, with some parts removed:

```py
class tUUID(BaseModel):
    i: Optional[str | UUID] = None
    uuid: UUID = Field(default_factory=uuid4)

    def __repr__(self):
        return self.uuid.hex

    def __str__(self):
        return self.uuid.hex

    def json(self):
        return str(self.uuid)

    def __eq__(self, other):
        return (self.uuid == other) or (str(self.uuid) == str(other)) or (self.uuid.hex == str(other))

    def __bool__(self) -> bool:
        # treat all 0s as the empty UUID
        return not (self.uuid.hex == '00000000000000000000000000000000')

    @root_validator(pre=True)
    @classmethod
    def coerce_uuid(cls, values):
        if not values.get('i', False):
            if not values.get('uuid', False):
                # TODO - does uuid4 wind up being called twice? things don't work w/o default_factory
                uuid_out = uuid4()
            else:
                uuid_out = values['uuid']
        else:
            val = values['i']
            if isinstance(val, UUID):
                uuid_out = val
            elif val == '':
                uuid_out = UUID('00000000000000000000000000000000')
            else:
                try:
                    uuid_out = UUID(val)
                except AttributeError as err:
                    logger.error(f'invalid uuid: {val}\n{traceback.format_exc()}')
                    raise err
        values = {
            'i': None,
            'uuid': uuid_out
        }
        # TODO - not sure why, but data structures aren't equalling right when i is different
        return values
```

My main question is there any way to get away from using two fields, one the input and one final value. If I get rid of the input, and do something like
 ```py
    @validator('uuid', pre=True)
    def coerce_uuid(cls, values) -> UUID:
        val = values.get('uuid')
        if val is None:
            # TODO - does uuid4 wind up being called twice? things don't work w/o default_factory
            uuid_out = uuid4()
        elif isinstance(val, UUID):
            uuid_out = val
        elif val == '':
            uuid_out = UUID('00000000000000000000000000000000')
        else:
            try:
                uuid_out = UUID(val)
            except AttributeError as err:
                logger.error(f'invalid uuid: {val}\n{traceback.format_exc()}')
                raise err
        return uuid_out
```

and try something like `testUUID(uuid="7e388947-7c9f-4d59-ab1d-ef3d7073b956")`

mypy complains: `error: Argument "uuid" to "testUUID" has incompatible type "str"; expected "UUID"  [arg-type]`

(One other oddity noted in the hint is that when I use tUUID in another pydantic model, say `foo` as `id`, if one is tUUID(i='the_uuid', uuid=UUID('the_uuid') and the other is on `bar`, `tUUID(i=None, uuid=UUID('the_UUID')), foo != bar when nothing is different between them but id, even though, given the __bool__ logic, foo.id == bar.id . To compensante, I've just set i to None in the validator.

Thanks, sorry for the long post, but I thought it would be clearest with all the code in question. Any advice is appreciated

**Reply by Samreay** [reply]
I'm back to hunting for solutions to this issue again - did you ever find a good way to solve this?

**Comment by rgiroud** [comment]
Hi, 
I have a "problem" that seems very similar (if not exactly?) this one that has been on my mind for a few months. 

I want to create a field in a model which accepts a set of types as input. The custom validator then does some logic based on the type of the input, while the output of the validator for that specific field is a very specific type. (What I'm talking about is more like a "Converter + Validator" rather than just a bare "Validator").

Here is a small reproducible example: Say I want a field that accept either a str or a float but outputs a str in any cases.

**Version 1:** 

```python
from pydantic import BaseModel, field_validator

class CustomModel(BaseModel): 
    a_string: str

    @field_validator("a_string", mode="before")
    @classmethod
    def validate_a_string(cls, input: str | float) -> str:
        return input if isinstance(input, str) else str(input)
```

![image](https://github.com/user-attachments/assets/13711a2b-fe9c-4ae9-bbf6-96f3219f407c)

This code has no runtime error. The static type hint of 'a_string' is effectively 'str', as wished. However, the static type hint when the input is a float raises a static error.

**Version 2:** 

```python
from pydantic import BaseModel, field_validator

class CustomModel(BaseModel): 
    a_string: str | float

    @field_validator("a_string", mode="before")
    @classmethod
    def validate_a_string(cls, input: str | float) -> str:
        return input if isinstance(input, str) else str(input)
```

![image](https://github.com/user-attachments/assets/dd58e187-7868-4e03-ae88-da627a0d79a2)

No runtime error. This time, the static type hinter doesn't raise any static error when the input is a float, as wished. However, the type hint of 'a_string' becomes 'str | float', although (by construction), at runtime 'a_string' will **always** be a 'str'. 

Is there any way to have the mix of the two desired behaviours? 

Thanks

**Reply by garbusbeach** [reply]
Yeah same issue for me. In my case it's a `datetime | str` input but casted into `datetime`, so it's especially frustrating with time format methods.

**Reply by petern3** [reply]
I've found a workaround for this whereby I overload the `__init__` function with the arguments and less strict type hints. It does remove a lot of the benefit of Pydantic validators, but it's still worth it for the rest of the benefit of Pydantic models. Something like:

```python
from pydantic import BaseModel

class CustomModel(BaseModel): 
    a_string: str
    
    def __init__(self, a_string: str | float, **kwargs):
        a_string_sanitized = a_string if isinstance(a_string, str) else str(a_string)
        super().__init__(a_string=a_string_sanitized, **kwargs)
```

<img width="426" height="191" alt="image" src="https://github.com/user-attachments/assets/8a282e8b-ca0b-49dd-836a-dcb53248be28" />

I still make use of validators (to validate), as long as they don't change the type.

**Comment by Jerry-Ma** [comment]
I am also facing this exact same issue. 
From the sounds of it, it seems like we should not change type of field in before validators. But I find in a lot of situations changing the type is part of the validation logic and it would be cumbersome to have a factory function that handles the type changing for the purpose of static type checking.

Is there a way we can provide additional "Input" annotations alongside the value annotation that they can be picked up by the static checker? For example:

```py
class MyModel(BaseModel):
    a: Annotated[str, InputType[str | float]] 
```

**Comment by s0hv** [comment]
The [ValidateAs](https://docs.pydantic.dev/latest/api/functional_validators/#pydantic.functional_validators.ValidateAs) annotation worked for my use case. Here is a example of converting a date string with a custom format to a `datetime`. This correctly annotates the `some_date` attribute only as a `datetime` while the input is validated as a string. I would imagine this approach can be used for at least some of the cases described in this discussion. It seems to work with a union input type as well, but my IDE complains about it.

```py
def parse_date(date_str: str) -> datetime:
    return datetime.strptime(date_str, '%Y.%m.%d %H:%M:%S').replace(tzinfo=tz.utc)

class MyModel(BaseModel):
    some_date: Annotated[datetime, ValidateAs(str, parse_date)]
```
